[Refactoring] Replace lifted breaks/returns with placeholder for async transform

hamishknight opened the pull request:
If we're lifting them outside of the control flow structure they're dealing with, turn them into placeholders, as they will no longer perform the control flow the user is expecting.

This handles:
- Return statements at the top-level of the callback.
- Break statements in switches that we re-write.

In addition, fix a bug where we'd inadvertently transform an unrelated switch statement in the callback.

Resolves rdar://74014897.

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4887,26 +4885,41 @@ class AsyncConverter : private SourceEntityWalker {
     return true;
   }
+
+  bool walkToStmtPre(Stmt *S) override {
```
What happens to returns/break inside closures?

hamishknight replied:
Updated to handle those cases

xymus commented:
@swift-ci Please smoke test macOS

hamishknight committed:
[Refactoring] Have addCustom take a SourceRange

And have it automatically track to the end of the
token, as that's the behavior we seem to always
want.

hamishknight committed:
[Refactoring] Don't transform unrelated switches

We were missing a `return` here to ignore any
switch statements that don't have anything to do
with the error handling.

hamishknight committed:
[Refactoring] Replace lifted breaks/returns with placeholders

If we're lifting them outside of the control flow
structure they're dealing with, turn them into
placeholders, as they will no longer perform the
control flow the user is expecting.

This handles:
- Return statements at the top-level of the callback.
- Break statements in switches that we re-write.

Resolves rdar://74014897.

hamishknight committed:
[Refactoring] [test] Use more strict matching on returns and breaks

Make sure we match against an exact return or
break in these cases, rather than a placeholder.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4680,6 +4686,7 @@ struct CallbackClassifier {
   void classifySwitch(SwitchStmt *SS) {
     if (!IsResultParam || singleSwitchSubject(SS) != ErrParam) {
       CurrentBlock->addNode(SS);
+      return;
```
Oof. Thanks for catching this and adding the test.

bnbarham reviewed (approved):
Looking good, thanks Hamish.
